src/Evolution/Systems/ScalarWave/Constraints.hpp

```diff
@@ -0,0 +1,105 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Evolution/Systems/ScalarWave/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+
+/// \cond
+namespace gsl {
+template <typename T>
+class not_null;
+}  // namespace gsl
+/// \endcond
+
+namespace ScalarWave {
+// @{
+/*!
+ * \brief Computes the scalar-wave one-index constraint.
+ *
+ * \details Computes the scalar-wave one-index constraint,
+ * \f$C_{i} = \partial_i\psi - \Phi_{i},\f$ which is
+ * given by Eq. (19) of \cite Holst2004wt
+ */
+template <size_t Dim>
+tnsr::i<DataVector, Dim, Frame::Inertial> one_index_constraint(
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& d_psi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi) noexcept;
+
+template <size_t Dim>
+void one_index_constraint(
+    gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*> constraint,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& d_psi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi) noexcept;
+// @}
+
+// @{
+/*!
+ * \brief Computes the scalar-wave 2-index constraint.
+ *
+ * \details Computes the scalar-wave 2-index FOSH constraint that
+ * [Eq. (20) of \cite Holst2004wt],
+ *
+ * \f{
+ * C_{ij} \equiv \partial_i \Phi_j - \partial_j \Phi_i
+ * \f}
+ *
+ * where \f$\Phi_{i} = \partial_i\psi\f$; and \f$\psi\f$ is the scalar field.
+ */
+template <size_t Dim>
+tnsr::ij<DataVector, Dim, Frame::Inertial> two_index_constraint(
+    const tnsr::ij<DataVector, Dim, Frame::Inertial>& d_phi) noexcept;
+
+template <size_t Dim>
+void two_index_constraint(
+    gsl::not_null<tnsr::ij<DataVector, Dim, Frame::Inertial>*> constraint,
+    const tnsr::ij<DataVector, Dim, Frame::Inertial>& d_phi) noexcept;
+// @}
+
+namespace Tags {
+/*!
+ * \brief Compute item to get the one-index constraint for the scalar-wave
+ * evolution system.
+ *
+ * \details See `one_index_constraint()`. Can be retrieved using
+ * `ScalarWave::Tags::OneIndexConstraint`.
+ */
+template <size_t Dim>
+struct OneIndexConstraintCompute : OneIndexConstraint<Dim>, db::ComputeTag {
+  using argument_tags =
+      tmpl::list<::Tags::deriv<Psi, tmpl::size_t<Dim>, Frame::Inertial>,
+                 Phi<Dim>>;
+  using return_type = tnsr::i<DataVector, Dim, Frame::Inertial>;
+  static constexpr void (*function)(
+      const gsl::not_null<return_type*> result,
+      const tnsr::i<DataVector, Dim, Frame::Inertial>&,
+      const tnsr::i<DataVector, Dim, Frame::Inertial>&) =
+      &one_index_constraint<Dim>;
```

Comment:
in case you end up deleting the return-by-value function you don't need the function arguments to disambiguate here anymore

Reply:
not deleting them